internal/receiver/smartagentreceiver/receiver.go

```diff
@@ -135,5 +148,79 @@ func (r *Receiver) createMonitor(monitorType string) (interface{}, error) {
 		output.AddExtraDimension(k, v)
 	}
 
-	return monitor, nil
+	// Configure SmartAgentConfigProvider to gather any global config overrides and
+	// set required envs.
+	configureEnvironmentOnce.Do(func() {
+		r.setUpSmartConfigProvider(extensions)
+		r.setUpEnvironment()
+	})
+
+	// Note, that this receiver has to configure main collectd even for collectd/custom monitor,
+	// despite the fact that, that monitor stands up its own instance of collectd to prevent this
+	// panic "Main collectd instance should not be accessed before being configured".
+	if r.config.monitorConfig.MonitorConfigCore().IsCollectdBased() {
+		configureCollectdOnce.Do(func() {
+			r.logger.Info("Configuring collectd")
+			err = configureMainCollectd(r.getCollectdConfig())
+		})
+	}
+
+	return monitor, err
+}
+
+func configureMainCollectd(collectdConfig *config.CollectdConfig) error {
+	return collectd.ConfigureMainCollectd(collectdConfig)
+}
+
+func (r *Receiver) setUpSmartConfigProvider(extensions map[configmodels.Extension]component.ServiceExtension) {
+	// If smartagent extension is not configured, use the default config.
+	f := smartagentextension.NewFactory()
+	defaultCfg := f.CreateDefaultConfig().(smartagentextension.SmartAgentConfigProvider)
+	r.config.saConfigProvider = defaultCfg
```

Comment:
Would it be better to make this a synchronized global instead of associated with the config for each receiver?  They aren't logically bound and config's are necessarily catch-alls.

Reply:
Yeah, that's a fair point. Updated per suggestion.